File: C++/17-树的子结构.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:
    bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2)
    {
        TreeNode *p1 = pRoot1, *p2 = pRoot2;
        if(!p1 || !p2)
            return false;
        bool result = false;
        if(p1->val == p2->val)
            result = helper(p1, p2);
        if(!result)
            result = HasSubtree(p1->left, p2);
        if(!result)
            result = HasSubtree(p1->right, p2);
        return result;
    }
    
    bool helper(TreeNode* p1, TreeNode* p2){
        if(!p2)
            return true;
        if(!p1)
            return false;
        if(p1->val == p2->val)
            return helper(p1->left, p2->left) and helper(p1->right, p2->right);
        return false;
    }
};

File: Python/17-树的子结构.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def HasSubtree(self, pRoot1, pRoot2):
        # write code here
        p1, p2 = pRoot1, pRoot2
        if not p1 or not p2:
            return False
        result = False
        if p1.val == p2.val:
            result = self.helper(p1, p2)
        if not result:
            result = self.HasSubtree(p1.left, p2)
        if not result:
            result = self.HasSubtree(p1.right, p2)
        return result
    
    def helper(self, p1, p2):
        if not p2:
            return True
        if not p1:
            return False
        if p1.val == p2.val:
            return self.helper(p1.left, p2.left) and self.helper(p1.right, p2.right)
        else:
            return False
